pkg/operator/replacement_starter.go

```diff
@@ -0,0 +1,323 @@
+package operator
+
+import (
+	"context"
+	"errors"
+	"os"
+	"reflect"
+	"time"
+
+	"k8s.io/client-go/dynamic/dynamicinformer"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned"
+	operatorclient "github.com/openshift/client-go/operator/clientset/versioned"
+	operatorinformer "github.com/openshift/client-go/operator/informers/externalversions"
+	"github.com/openshift/library-go/pkg/controller/controllercmd"
+	libgoetcd "github.com/openshift/library-go/pkg/operator/configobserver/etcd"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/genericoperatorclient"
+	"github.com/openshift/library-go/pkg/operator/resourcesynccontroller"
+	"github.com/openshift/library-go/pkg/operator/status"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"github.com/openshift/multi-operator-manager/pkg/library/libraryapplyconfiguration"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/clock"
+)
+
+type AuthenticationOperatorInput struct {
+	kubeClient                   kubernetes.Interface
+	configClient                 configclient.Interface
+	operatorClient               operatorclient.Interface
+	authenticationOperatorClient v1helpers.OperatorClient
+	eventRecorder                events.Recorder
+
+	informers []SimplifiedInformerFactory
+}
+
+type SimplifiedInformerFactory interface {
+	Start(ctx context.Context)
+	WaitForCacheSync(ctx context.Context)
+}
+
+type OperatorStarter interface {
+	RunOnce(ctx context.Context) error
+	Start(ctx context.Context) error
+}
+
+type authenticationOperatorStarter struct {
+	eventRecorder events.Recorder
+
+	informers        []SimplifiedInformerFactory
+	controllerRunFns []func(ctx context.Context, workers int)
+	// TODO make a function specifically for syncOnce, this looks fragile
+	controllerRunOnceFns []func(ctx context.Context, syncCtx factory.SyncContext) error
+}
+
+func (a authenticationOperatorStarter) RunOnce(ctx context.Context) error {
+	if err := a.startInformers(ctx); err != nil {
+		return err
+	}
+
+	errs := []error{}
+
+	for _, controllerRunOnceFn := range a.controllerRunOnceFns {
+		// TODO add timeout.
+		syncCtx := factory.NewSyncContext("run-once-sync-context", a.eventRecorder)
+		err := controllerRunOnceFn(ctx, syncCtx)
+		errs = append(errs, err)
+	}
+	return errors.Join(errs...)
+}
+
+func (a authenticationOperatorStarter) startInformers(ctx context.Context) error {
```

Comment:
change the signature of this method, this method doesn't return any err

Reply:
> change the signature of this method, this method doesn't return any err

This one I think we want to keep to allow flexibility for operator authors